tests/test_validators_dataclass.py

```diff
@@ -0,0 +1,600 @@
+    assert MyDataclass(a=False).a == 0
+    assert MyDataclass(a=4.5).a == 4
+
+
+def test_validate_whole():
+    @dataclass
+    class MyDataclass:
+        a: List[int]
+
+        @validator('a', whole=True, pre=True)
+        def check_a1(cls, v):
+            v.append('123')
+            return v
+
+        @validator('a', whole=True)
+        def check_a2(cls, v):
+            v.append(456)
+            return v
+
+    assert MyDataclass(a=[1, 2]).a == [1, 2, 123, 456]
+
+
+def test_validate_kwargs():
+    @dataclass
+    class MyDataclass:
+        b: int
+        a: List[int]
+
+        @validator('a')
+        def check_a1(cls, v, values, **kwargs):
+            return v + values['b']
+
+    assert MyDataclass(a=[1, 2], b=6).a == [7, 8]
+
+
+def test_validate_whole_error():
+    calls = []
+
+    @dataclass
+    class MyDataclass:
+        a: List[int]
+
+        @validator('a', whole=True, pre=True)
+        def check_a1(cls, v):
+            calls.append(f'check_a1 {v}')
+            if 1 in v:
+                raise ValueError('a1 broken')
+            v[0] += 1
+            return v
+
+        @validator('a', whole=True)
+        def check_a2(cls, v):
+            calls.append(f'check_a2 {v}')
+            if 10 in v:
+                raise ValueError('a2 broken')
+            return v
+
+    assert MyDataclass(a=[3, 8]).a == [4, 8]
+    assert calls == ['check_a1 [3, 8]', 'check_a2 [4, 8]']
+
+    calls = []
+    with pytest.raises(ValidationError) as exc_info:
+        MyDataclass(a=[1, 3])
+    assert exc_info.value.errors() == [{'loc': ('a',), 'msg': 'a1 broken', 'type': 'value_error'}]
+    assert calls == ['check_a1 [1, 3]']
+
+    calls = []
+    with pytest.raises(ValidationError) as exc_info:
+        MyDataclass(a=[5, 10])
+    assert exc_info.value.errors() == [{'loc': ('a',), 'msg': 'a2 broken', 'type': 'value_error'}]
+    assert calls == ['check_a1 [5, 10]', 'check_a2 [6, 10]']
+
+
+class Config:
+    validate_assignment = True
+
+
+@dataclass(config=Config)
+class ValidateAssignmentDataclass:
+    a: int = 4
+    b: str = ...
+
+    @validator('b')
+    def b_length(cls, v, values, **kwargs):
+        if 'a' in values and len(v) < values['a']:
+            raise ValueError('b too short')
+        return v
+
+
+def test_validating_assignment_ok():
+    p = ValidateAssignmentDataclass(b='hello')
+    assert p.b == 'hello'
+
+
+def test_validating_assignment_fail():
+    with pytest.raises(ValidationError):
+        ValidateAssignmentDataclass(a=10, b='hello')
+
+    p = ValidateAssignmentDataclass(b='hello')
+    with pytest.raises(ValidationError):
+        p.b = 'x'
+
+
+def test_validating_assignment_dict():
+    with pytest.raises(ValidationError) as exc_info:
+        ValidateAssignmentDataclass(a='x', b='xx')
+    assert exc_info.value.errors() == [
+        {'loc': ('a',), 'msg': 'value is not a valid integer', 'type': 'type_error.integer'}
+    ]
+
+
+def test_validate_multiple():
```

Comment:
We don't need all these tests copied from `test_validators`, just the critical tests to check dataclass validators work.

Reply:
Chose what I think are the most important ones, feel free to make suggestions!